### DISCUSSION TITLE
How to use Wagmi with Ethers Js

### DISCUSSION THREAD

**Discussion by cryptobredda** [wevm/wagmi#2565] [Q&A]
Hi, 

I would like to convert the veim backend to a ethers Js version instead.

This means I would like to be able to use provider and signer instead of using hooks. I am upgrading my app and it has tons of features that rely on ether and cant be change to the methodologies of wagmi. so this is important for me. 

The reason why I am using wagmi is because my wallet connector app is using wagmi so I am kinda forced. 

I hoep you understand my request and be able to help me.

Goal:

To have a hook / constant that simulates the usage of provider and signer from the ether library.

example:

useSigner()
useProvider()

and then I can access ethers functiosn from those.

**Comment by jxom** [comment]
https://wagmi.sh/react/ethers-adapters

**Reply by cryptobredda** [reply]
I have read this before making this ticket.

I am having issues with it as its not converting the provider and the signer correctly.

example, 

using the provider to get the signer directly gives an error.

additionally the signer hook doesnt work at all. it keeps saying undefined. thus I made this ticket

**Comment by cryptobredda** [comment]
Additionally there is anthoer problem I am facing. 

when changing nextwork from goeril to any other network, the signer break and gives an error about the chain being undefined.
This is switching from Goeril to Mumbai


![image](https://github.com/wagmi-dev/wagmi/assets/36314021/a1358fdc-0acb-4c74-8cc1-b60bb34d9154)

**Comment by cryptobredda** [comment]
@jxom  What do you think this could be?

**Comment by cryptobredda** [comment]
@jxom @riklomas @bpierre @challet 

I am still running into an issue, could you please assist me with it?

Further information to help you.

I used the code in the Ether Adapter docs but I am running into an issue when switching networks. It seems like that chainId is defined only for Etheruem and Goeril but not other chains. How do I get it to work for other chains as well.

**Comment by artschekoff** [comment]
where is ethers adaptor? 404...

I still need to have ethers with wagmi, for the damn god, I don't want to switch to viem or how does this called...

**Reply by artschekoff** [reply]
Upd - new link is here - https://wagmi.sh/core/guides/ethers#ethers-js-adapters

**Reply by L1xus** [reply]
Hi sir!
I'm running into the same issue!
did you fix it ?

**Reply by artschekoff** [reply]
Hello! Yes, but my case is more complicated. I also have web3modal and walletconnect in my project, and, unfortunately, I can't use it with last version of wagmi. 

So I have to downgrade wagmi to lower version, as well as wevm, and also I found archived version of ethers adapters for this one (listed down below)

Hope this will help to you:

**package.json**

`    "viem": "1.21.4",
    "wagmi": "1.4.13",
    "@wagmi/connectors": "3.1.11",
    "@wagmi/core": "1.4.13",
    "@web3modal/wagmi": "^3.5.6",
`

**ethers.ts**

```
import { WalletClient, getPublicClient, getWalletClient } from '@wagmi/core'
import { BrowserProvider, FallbackProvider, JsonRpcProvider, JsonRpcSigner } from 'ethers'
import { type Chain, type Client, type Transport } from 'viem'

export function clientToProvider(client: Client<Transport, Chain>) {
  const { chain, transport } = client
  const network = {
    chainId: chain.id,
    name: chain.name,
    ensAddress: chain.contracts?.ensRegistry?.address,
  }
  if (transport.type === 'fallback') {
    const providers = (transport.transports as ReturnType<Transport>[]).map(
      ({ value }) => new JsonRpcProvider(value?.url, network)
    )
    if (providers.length === 1) return providers[0]
    return new FallbackProvider(providers)
  }
  return new JsonRpcProvider(transport.url, network)
}

/** Action to convert a viem Client to an ethers.js Provider. */
export function getEthersProvider({ chainId }: { chainId?: number } = {}) {
  const client = getPublicClient({ chainId })
  return clientToProvider(client)
}

export function walletClientToSigner(walletClient: WalletClient) {
  const { account, chain, transport } = walletClient
  const network = {
    chainId: chain.id,
    name: chain.name,
    ensAddress: chain.contracts?.ensRegistry?.address,
  }
  const provider = new BrowserProvider(transport, network)
  const signer = new JsonRpcSigner(provider, account.address)
  return signer
}

/** Action to convert a viem Wallet Client to an ethers.js Signer. */
export async function getEthersSigner({ chainId }: { chainId?: number } = {}) {
  const client = await getWalletClient({ chainId })
  if (!client) return undefined
  return walletClientToSigner(client)
}
```

**custom useProvider.ts**

```
import { FallbackProvider, JsonRpcProvider } from 'ethers'
import { useRef } from 'react'
import { getEthersProvider } from './ethers'

export type Web3Provider = JsonRpcProvider | FallbackProvider

type S = Web3Provider

type P = {
  chainId?: number
}

const useProvider = ({ chainId }: P = {}) => useRef<S>(getEthersProvider({ chainId })).current

export default useProvider
```

**custom useSigner.ts**
```
import { JsonRpcSigner } from 'ethers'
import { useEffect, useState } from 'react'
import { useAccount, useNetwork } from 'wagmi'
import { getEthersSigner } from './ethers'

type S = {
  isLoading: boolean
  isFetched: boolean
  data: JsonRpcSigner | null
}
type P = {
  chainId?: number
}
const useSigner = ({ chainId: _chainId }: P = {}) => {
  const { address } = useAccount()
  const { chain } = useNetwork()

  const [s, setS] = useState<S>({
    isLoading: false,
    isFetched: false,
    data: null,
  })

  useEffect(() => {
    const getSigner = async () => {
      setS((x) => ({ ...x, isLoading: true }))

      const signer = await getEthersSigner({ chainId: _chainId ?? chain?.id })

      setS((x) => ({ ...x, isLoading: false, data: signer ?? null, isFetched: !!signer }))
    }

    if (address) {
      getSigner()
    }
  }, [address])

  return s
}

export default useSigner
```

**Reply by L1xus** [reply]
Thanks for your help!
I fixed the issue!
I'm using walletconnect as well and didn't know that it's not supporting wagmi latest version so I was following the wagmi v2 guide and getting conflicts with walletConnect configs :)
now I have new problems I have to solve lol

**Reply by artschekoff** [reply]
maybe this will help)) my web3modal injection with versions above:

```
'use client'

import { walletConnectProvider } from '@web3modal/wagmi'
import { WagmiConfig, configureChains, createConfig } from 'wagmi'
import { CoinbaseWalletConnector } from 'wagmi/connectors/coinbaseWallet'
import { InjectedConnector } from 'wagmi/connectors/injected'
import { WalletConnectConnector } from 'wagmi/connectors/walletConnect'
import { publicProvider } from 'wagmi/providers/public'

import { dappConfig } from '~/config'

type Web3ModalContextProps = {
  children: React.ReactNode
}

export const projectId = process.env.NEXT_PUBLIC_WALLETCONNECT_ID!

const { chains, publicClient } = configureChains(dappConfig.chains, [
  walletConnectProvider({ projectId }),
  publicProvider(),
])

const metadata = {
  name: 'Web3Modal',
  description: 'Web3Modal Example',
  url: 'https://web3modal.com',
  icons: ['https://avatars.githubusercontent.com/u/37784886'],
}

const wagmiConfig = createConfig({
  autoConnect: true,
  connectors: [
    new WalletConnectConnector({ chains, options: { projectId, showQrModal: true, metadata } }),
    new InjectedConnector({ chains, options: { shimDisconnect: true } }),
    new CoinbaseWalletConnector({ chains, options: { appName: metadata.name } }),
  ],
  publicClient,
})

export const Web3ModalInjection: React.FC<Web3ModalContextProps> = ({ children }) => (
  <>
    <WagmiConfig config={wagmiConfig}>{children}</WagmiConfig>
  </>
)

```
